zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZNodeGroupAclProperties.java

```diff
@@ -0,0 +1,114 @@
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Configured properties for ZNode Group ACL feature
+ */
+public class ZNodeGroupAclProperties {
+  private static ZNodeGroupAclProperties instance = null;
+
+  private ZNodeGroupAclProperties() {
+  }
+
+  public static ZNodeGroupAclProperties getInstance() {
+    if (instance == null) {
+      synchronized (ZNodeGroupAclProperties.class) {
+        if (instance == null) {
+          instance = new ZNodeGroupAclProperties();
+        }
+      }
+    }
+    return instance;
+  }
+
+  /**
+   * Property key values (JVM configs) for ZNode group ACL features
+   */
+  public static final String ZNODE_GROUP_ACL_CONFIG_PREFIX = "zookeeper.ssl.znodeGroupAcl.";
+  // Enables/disables whether znodes created by auth'ed clients
+  // should have ACL fields populated with the client Id given by the authentication provider.
+  // Has the same effect as the ZK client using ZooDefs.Ids.CREATOR_ALL_ACL.
+  public static final String SET_X509_CLIENT_ID_AS_ACL =
+      ZNODE_GROUP_ACL_CONFIG_PREFIX + "setX509ClientIdAsAcl";
+  // A list of domain names that will have super user privilege, separated by ","
+  private static final String SUPER_USER_DOMAIN_NAME =
+      ZNODE_GROUP_ACL_CONFIG_PREFIX + "superUserDomainName";
+  // A list of znode path prefixes, separated by ","
+  // Znode whose path starts with the defined path prefix would have open read access
+  // Meaning the znode will have (world:anyone, r) ACL
+  private static final String OPEN_READ_ACCESS_PATH_PREFIX =
+      ZNODE_GROUP_ACL_CONFIG_PREFIX + "openReadAccessPathPrefix";
+  private Set<String> openReadAccessPathPrefixes;
+  private Set<String> superUserDomainNames;
+
+  /**
+   * Get open read access path prefixes from config
+   * @return A set of path prefixes
+   */
+  private Set<String> loadOpenReadAccessPathPrefixes() {
+    String openReadAccessPathPrefixesStr = System.getProperty(OPEN_READ_ACCESS_PATH_PREFIX);
+    if (openReadAccessPathPrefixesStr == null || openReadAccessPathPrefixesStr.isEmpty()) {
+      return Collections.emptySet();
+    }
+    return Arrays.stream(openReadAccessPathPrefixesStr.split(",")).filter(str -> str.length() > 0)
+        .collect(Collectors.toSet());
+  }
+
+  /**
+   * Get the domain names that are mapped to super user access privilege
+   * @return A set of domain names
+   */
+  private Set<String> loadSuperUserDomainNames() {
+    String superUserDomainNameStr = System.getProperty(SUPER_USER_DOMAIN_NAME);
+    if (superUserDomainNameStr == null || superUserDomainNameStr.isEmpty()) {
+      return Collections.emptySet();
+    }
+    return Arrays.stream(superUserDomainNameStr.split(",")).filter(str -> str.length() > 0)
+        .collect(Collectors.toSet());
+  }
+
+  public Set<String> getOpenReadAccessPathPrefixes() {
+    if (openReadAccessPathPrefixes == null) {
+      synchronized (ZNodeGroupAclProperties.class) {
```

Comment:
synchronize on `openReadAccessPathPrefixes`

Reply:
I did this because `openReadAccessPathPrefixes` is a non-final object, it is not a good practice to synchronize on a non-final object.

Reply:
agreed, missed it not being `final`. the concern is the scope of synchronization here, but I don't think it's a big deal. ideally we'd be creating a final lock for each set, but I think it's fine either way.